test/route-recognizer.spec.js

```diff
@@ -1,8 +1,136 @@
+    isDynamic: true,
+    path: '/dynamic/foo/star/test/path',
+    params: { id: 'foo', path: 'test/path' }
+  }];
+
 describe('route recognizer', () => {
-  it('should have some tests', () => {
-    var recognizer = new RouteRecognizer();
-    expect(recognizer).not.toBe(null);
+  it('should reject unknown routes', () => {
+    let recognizer = new RouteRecognizer();
+
+    expect(recognizer.hasRoute('static')).toBe(false);
+    expect(() => recognizer.handlersFor('static')).toThrow();
+    expect(() => recognizer.generate('static')).toThrow();
+    expect(recognizer.recognize('/notfound')).toBeUndefined();
+  });
+
+  it('should register unnamed routes', () => {
+    let recognizer = new RouteRecognizer();
+    recognizer.add([{'path': 'b','handler': {}}]);
+
+    expect(recognizer.names).toEqual({});
+    expect(recognizer.recognize('/b')).toBeTruthy();
+  });
+
+  for (let routeTest of routeTestData) {
+    it(`should recognize ${routeTest.title}`, () => {
+      let recognizer = new RouteRecognizer();
+      recognizer.add([routeTest.route]);
+
+      let result = recognizer.recognize(routeTest.path);
+      expect(result).toBeTruthy();
+      expect(result.length).toBe(1);
+      expect(result[0].handler).toEqual(routeTest.route.handler);
+      expect(result[0].isDynamic).toBe(routeTest.isDynamic);
+      expect(result[0].params).toEqual(routeTest.params);
+    });
+
+    it(`should generate ${routeTest.title}`, () => {
+      let recognizer = new RouteRecognizer();
+      recognizer.add([routeTest.route]);
+
+      expect(recognizer.generate(routeTest.route.handler.name, routeTest.params))
+        .toBe(routeTest.path);
+    });
+  }
+
+  it('should find handlers by route name', () => {
+    let recognizer = new RouteRecognizer();
+    recognizer.add([staticRoute]);
+
+    expect(recognizer.hasRoute('static')).toBe(true);
+    expect(recognizer.handlersFor('static')[0].handler).toEqual(staticRoute.handler);
+  });
+
+  it('should generate query strings', () => {
+    let recognizer = new RouteRecognizer();
+    let gen = recognizer.generateQueryString.bind(recognizer);
+
+    expect(gen()).toBe('');
+    expect(gen(null)).toBe('');
+    expect(gen({})).toBe('');
+    expect(gen({ a: null })).toBe('');
+
+    expect(gen({ '': 'a' })).toBe('?=a');
+    expect(gen({ a: 'b' })).toBe('?a=b');
+    expect(gen({ a: 'b', c: 'd' })).toBe('?a=b&c=d');
+    expect(gen({ a: 'b', c: null })).toBe('?a=b');
+
+    expect(gen({ a: [ 'b', 'c' ]})).toBe('?a[]=b&a[]=c');
+    expect(gen({ '&': [ 'b', 'c' ]})).toBe('?%26[]=b&%26[]=c');
+
+    expect(gen({ a: '&'})).toBe('?a=%26');
+    expect(gen({ '&': 'a'})).toBe('?%26=a');
+    expect(gen({ a: true})).toBe('?a=true');
+  });
+
+  it('should parse query strings', () => {
+    let recognizer = new RouteRecognizer();
+    let parse = recognizer.parseQueryString.bind(recognizer);
+
+    expect(parse('')).toEqual({});
+    expect(parse('=')).toEqual({});
```

Comment:
Some other query string parsers get `{ "": "" }` here. I'm not sure if that's more correct, but it doesn't seem useful.